package/src/index/mod.rs

```diff
@@ -0,0 +1,514 @@
+
+            let tmp_dir = tmp_dir.into_path();
+            std::fs::rename(tmp_dir, target_dir).with_path(target_dir)?;
+        }
+
+        Ok(())
+    }
+}
+
+impl PackageIndex<Exclusive> {
+    /// Opens the package index for writing.
+    ///
+    /// If the package index doesn't exist, creates an empty one.
+    pub fn exclusive(config: Config) -> Result<Self, Error> {
+        let lock = IndexLock::exclusive(&config)?;
+        if !config.index_dir.exists() {
+            std::fs::create_dir_all(&config.index_dir).with_path(&config.index_dir)?;
+        }
+        Ok(PackageIndex {
+            cache: RefCell::new(PackageIndexCache {
+                config,
+                _lock: lock,
+                package_files: HashMap::new(),
+            }),
+        })
+    }
+
+    /// Writes a package to the index.
+    pub fn save(&mut self, pkg: Package) -> Result<(), Error> {
+        self.cache.borrow_mut().save(pkg)
+    }
+}
+
+/// The identifier of a package in the package index.
+#[derive(Clone, PartialEq, Eq, Debug, Hash, Deserialize, PartialOrd, Ord)]
+pub enum Id {
+    Github { org: String, name: String },
+}
+
+impl Id {
+    /// Returns the path (relative to the package index base directory) where this
+    /// package should be stored.
+    pub fn path(&self) -> PathBuf {
+        match self {
+            Id::Github { org, name } => PathBuf::from(format!("github/{org}/{name}")),
+        }
+    }
+
+    pub fn remote_url(&self) -> Result<gix::Url, Error> {
+        match self {
+            Id::Github { org, name } => Ok(format!("https://github.com/{org}/{name}").try_into()?),
+        }
+    }
+}
+
+impl std::fmt::Display for Id {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Id::Github { org, name } => write!(f, "github/{org}/{name}"),
+        }
+    }
+}
+
+#[derive(Debug)]
+pub enum IdParseError {
+    /// We expect exactly 2 slashes, and return this error if there aren't.
+    Slashes,
+    /// We only know about github right now, and return this error if they ask for a different one.
+    UnknownIndex { index: String },
+    /// Our rules for user and package names are currently the same as Nickel's identifier rules.
+    InvalidId { id: String },
+}
+
+impl std::error::Error for IdParseError {}
+
+impl std::fmt::Display for IdParseError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            IdParseError::Slashes => {
+                write!(f, "doesn't match the expected <index>/<org>/<name> pattern")
+            }
+            IdParseError::UnknownIndex { index } => write!(
+                f,
+                "unknown index `{index}`, the only valid value is `github`"
+            ),
+            IdParseError::InvalidId { id } => write!(f, "invalid identifier `{id}`"),
+        }
+    }
+}
+
+static ID_REGEX: LazyLock<Regex> =
```

Comment:
I didn't know something like `LazyLock` existed in the standard library. I wonder if this deprecates the usage we make of `lazy_static` elsewhere in Nickel.

Reply:
Yes, I think so. It's only been in the stdlib since 1.80